### DISCUSSION TITLE
Web search

### DISCUSSION THREAD

**Discussion by mattlindsey** [AllYourBot/hostedgpt#558] [General]
I don't use ChatGPT much, but I was recently pleasantly surprised when I asked it about 'recent events' and it utilized (new?) web search capabilities to give a great answer:
https://openai.com/index/introducing-chatgpt-search/

I think adding something like that to HostedGPT would be a nice selling point.

**Comment by keithschacht** [comment]
I too would like this feature! Code wise, I don’t think it’s too much to implement. It would basically mean copying the Memory Toolbox class and calling it Browser (or something similar). Obviously changing the prompt to explain “if the user asks you to look something up or if they’re asking questions about current information which are after your knowledge cutoff then …” and then just implementing a sever-side HTTP request to google.com. Maybe the search results would need to be parsed to turn into some simplified markdown.

I think the two tool calls we’d need to support would be something like:

search_google
visit_url

Want to take a first pass at a PR?

**Reply by mattlindsey** [reply]
Sure. I'll try to have something this week or next.

**Reply by mattlindsey** [reply]
Wouldn't issuing a request to google.com give results that are really hard to parse?  I think I'll need https://developers.google.com/custom-search/v1/overview or https://serpapi.com/search-api.  They're both free for 100 searches a month, but google is much cheaper after that.  With either of those, the user will need to get an API key.

Or are you saying the Browser tool would just provide the html returned from google.com, converted to markdown using a tool like https://pandoc.org?  I've seen pandoc used elsewhere but it would need to be installed by the user.  Or just parsed a little and reformatted with Nokogiri?

I guess I'm a little confused how this would work exactly.

**Reply by keithschacht** [reply]
I've used serpapi in the past for other projects. It is a nice API.  However, I was thinking that it wouldn't be too hard to parse google's results. Although, I haven't actually investigated that closely. But it would be really nice if we could make the search tool "just work" without everyone having to create an account on another service and provide additional configuration to the app.

I just asked a chatgpt these questions and it provided some interesting suggestions. Here is the full conversation (it's pretty short) but the notable takeaways were:

1. Google supports adding `&tbs=li:1` to their search results which simplifies the resulting page a decent amount. I like that this eliminates their Google AI result from the page, but I don't like that it eliminates the Google Onebox answer since that sometimes contains useful information.
2. ChatGPT wrote a couple examples of ruby code that would strip only the user visible text from the resulting page and/or would convert the resulting page to markdown

I suspect that one of these could work.

conversation here: https://chatgpt.com/share/674345ad-2e48-800a-aa61-10a2cbe4e799

A v1 of this feature could be: we create a tool that tells the LLM it has the ability to search google for any questions where it thinks searching google would help it answer, and we perform a google search and simply dump all the text from the search results page back into the conversation history. This means the LLM doesn't have the ability to click any of the links and further navigate to the page, but often you can glean quite a bit of useful information just from the google search results page (page titles and page snippets).

**Reply by keithschacht** [reply]
(Notably I had to go to chatgpt in order to have a conversation I could share with you. At some point we should support creating share links like this within HostedGPT! :) )

**Reply by mattlindsey** [reply]
I'll try to do a basic `search_google(query:)` tool like you suggest which just dumps the scraped text from the results page (probably as markdown).  

I'm worried it will dump a bunch of junk in there, but it's certainly worth a try.   Actually it may work well if we tell the LLM in the tool instructions to use a good detailed query string.

**Comment by mattlindsey** [comment]
`google_search` tool call was implemented in #567 that returns scraped text from www.google.com/search (including anchor links in parens).

`visit_url` was mentioned as next tool call to implement.  Is that still necessary, or what should we implement next?  Maybe after a google_search is done the llm can decide if it wants to do a `visit_url` on one or more of the links?  If so, how would that work?

**Comment by mattlindsey** [comment]
#731
